src/librustc/ty/sty.rs

```diff
@@ -879,28 +879,58 @@ pub enum InferTy {
 #[derive(Clone, Copy, PartialEq, Eq, Hash, Debug, RustcEncodable, RustcDecodable)]
 pub struct ExistentialProjection<'tcx> {
     // FIXME(tschottdorf): give this the same treatment as #42171 (in this PR).
-    pub trait_ref: ExistentialTraitRef<'tcx>,
-    pub item_name: Name,
+    //pub trait_ref: ExistentialTraitRef<'tcx>,
+    //pub item_name: Name,
+    pub item_def_id: DefId,
+    pub substs: &'tcx Substs<'tcx>,
     pub ty: Ty<'tcx>,
 }
 
-pub type PolyExistentialProjection<'tcx> = Binder<ExistentialProjection<'tcx>>;
+impl<'tcx> ExistentialProjection<'tcx> {
+    pub fn from_ty_ref_and_name(
+        tcx: TyCtxt, ty: Ty<'tcx>, trait_ref: ty::ExistentialTraitRef<'tcx>, item_name: Name,
+    ) -> ExistentialProjection<'tcx> {
+        let item_def_id = tcx.associated_items(trait_ref.def_id).find(
+            |item| item.name == item_name).unwrap().def_id;
 
-impl<'a, 'tcx, 'gcx> ExistentialProjection<'tcx> {
-    pub fn item_name(&self) -> Name {
-        self.item_name // safe to skip the binder to access a name
+        ExistentialProjection {
+            ty: ty,
+            item_def_id: item_def_id,
+            substs: trait_ref.substs,
+        }
     }
 
+    pub fn item_name(&self, tcx: TyCtxt) -> Name {
+        tcx.associated_item(self.item_def_id).name
+    }
+
+    pub fn trait_ref(&self, tcx: TyCtxt) -> ty::ExistentialTraitRef<'tcx> {
+        let def_id = match tcx.associated_item(self.item_def_id).container {
+            ty::TraitContainer(def_id) => def_id,
+            ty::ImplContainer(def_id) => {
+                bug!("asked for a TraitContainer, got an ImplContainer with DefId {:?}", def_id)
+            }
+        };
+        ty::ExistentialTraitRef{
+            def_id: def_id,
+            substs: self.substs,
+        }
+    }
+}
+
+pub type PolyExistentialProjection<'tcx> = Binder<ExistentialProjection<'tcx>>;
+
+impl<'a, 'tcx, 'gcx> ExistentialProjection<'tcx> {
     pub fn sort_key(&self, tcx: TyCtxt<'a, 'gcx, 'tcx>) -> (ich::Fingerprint, InternedString) {
         // We want something here that is stable across crate boundaries.
         // The DefId isn't but the `deterministic_hash` of the corresponding
         // DefPath is.
-        let trait_def = tcx.trait_def(self.trait_ref.def_id);
+        let trait_def = tcx.trait_def(self.trait_ref(tcx).def_id);
```

Comment:
@michaelwoerister Given that we have the `DefId` of the associated type, can't we skip all of this stuff and use that directly?